File: programmers/PART12/올바른괄호.js
// https://programmers.co.kr/learn/courses/30/lessons/12909?language=javascript

// url

/**
 * s = (())()
 *
 * s를 stack에 차례차례 push한다. 이 때, stack의 맨 상단을 확인하여 매칭되는 괄호를 만난다면 해당 괄호를 pop한다.
 *    ch
 * 1: "(" -> ["("] <- push "("
 * 2: "(" -> ["(", "("] <- push "("
 * 3: ")" -> ["("] -> pop "("
 * 4: ")" -> [] -> pop "("
 * 5: "(" -> ["("] <- push
 * 6: ")" -> [] -> pop
 *
 * 마지막에 stack이 비어있다면 올바른 괄호라고 할 수 있다.
 *
 * s = )()()(
 *    ch
 * 1: ")" -> [")"] <- push ")"
 * 2: "(" -> [")", "("] <- push "("
 * 3: ")" -> [")"] -> pop "("
 * 4: "(" -> [")", "("] <- push "("
 * 5: ")" -> [")"] -> pop "("
 * 6: "(" -> [")", "("] <- push "("
 *
 * stack이 비어있지 않으므로 올바른 괄호가 아니다.
 */
function solution(s) {
  const stack = [];
  const parens = { ")": "(" };
  const top = () => stack[stack.length - 1];
  const match = (c) => !!top() && parens[c] === top();
  for (const ch of s) {
    if (!match(ch)) {
      stack.push(ch);
    } else {
      stack.pop();
    }
  }
  return !stack.length;
}

let s;
// s = "()()";
// console.log(solution(s));
s = "(())()";
console.log(solution(s));
// s = ")()()(";
// console.log(solution(s));
// s = "(()(";
// console.log(solution(s));

File: programmers/PART12/올바른괄호.py
# https://programmers.co.kr/learn/courses/30/lessons/12909

def solution(s):
    """
    stack 자료 구조를 사용한다.

    s = "()()"

    1: ( => [ ( ]: stack의 대치되는 괄호가 없으므로 ( 를 그대로 넣는다.
    2: ) => [ ]: stack의 top에 대치되는 괄호가 있으므로 top을 제거한다.

    stack이 비었으므로 true를 반환한다.
    """

    stack = list()
    paren = {")": "("}

    for c in s:
        if len(stack) and stack[-1] == paren.get(c):
            stack.pop()
        else:
            stack.append(c)
    return not bool(len(stack))


s = "()()"
print(solution(s))  # true
s = "(())()"
print(solution(s))  # true
s = ")()("
print(solution(s))  # false
s = "(()("
print(solution(s))  # false
